src/classes/TemporaryFile.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  S3Client,
+  CreateMultipartUploadCommand,
+  UploadPartCommand,
+  CompleteMultipartUploadCommand,
+} from '@aws-sdk/client-s3';
+import { v4 as uuid } from 'uuid';
+import { SystemConfigurationError } from '../errors';
+import type {
+  CompletedPart,
+} from '@aws-sdk/client-s3';
+
+export class TemporaryFile {
+  public readonly permanentFileSystemPath: string;
+
+  private uploadedLength = 0;
+
+  private s3Url?: string;
+
+  private uploadId?: string;
+
+  private readonly bucket: string;
+
+  private readonly key: string;
+
+  private partNumber = 1;
+
+  private readonly s3: S3Client;
+
+  private readonly completedParts: CompletedPart[] = [];
+
+  public constructor(permanentFileSystemPath: string) {
+    if (process.env.TEMPORARY_FILE_S3_BUCKET === undefined) {
+      throw new SystemConfigurationError('TEMPORARY_FILE_S3_BUCKET must be populated in order to upload to s3.');
+    }
+    if (process.env.AWS_ACCESS_KEY_ID === undefined) {
+      throw new SystemConfigurationError('AWS_ACCESS_KEY_ID must be populated in order to upload to s3.');
+    }
+    if (process.env.AWS_ACCESS_SECRET === undefined) {
+      throw new SystemConfigurationError('AWS_ACCESS_SECRET must be populated in order to upload to s3.');
+    }
+    if (process.env.TEMPORARY_FILE_S3_BUCKET_REGION === undefined) {
+      throw new SystemConfigurationError('TEMPORARY_FILE_S3_BUCKET_REGION must be populated in order to upload to s3.');
+    }
```

Comment:
If this is how all files are going to be uploaded, would it make sense to check for these env vars at startup rather than in this constructor? If a missing env var means the service won't work at all I'd prefer to learn about that as soon as possible.

Reply:
Great idea! That said, I would also need / want to also check in the constructor for defensiveness and type safety purposes (this class can't assume that things were checked in an unrelated scope)